core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -4049,16 +4049,23 @@ object Stream extends StreamLowPriority {
                       .compile
                       .drain
                       .guaranteeCase(oc =>
-                        lease.cancel
-                          .flatMap(onOutcome(oc, _)) >> available.release >> decrementRunning
+                        lease.cancel.rethrow
+                          .guaranteeCase {
+                            case Outcome.Succeeded(fu) =>
+                              onOutcome(oc <* Outcome.succeeded(fu), Either.unit)
+
+                            case Outcome.Errored(e) =>
+                              onOutcome(oc, Either.left(e))
+
+                            case _ =>
+                              F.unit
+                          }
+                          .forceR(available.release >> decrementRunning)
```

Comment:
I think this is correct but slightly hard to reason about. Can it not be moved inside the `guaranteeCase`?

Reply:
Agree, moved out

Reply:
I actually meant something else, can it not be moved inside the guaranteeCase so that `forceR` is not needed at all?

Reply:
So same order as now, but instead of `forceR` a `flatMap`? No, because stream can be compiled with monad error. If this is still not what you mean, please provide code snippet.

Reply:
What I mean is that now the `release` is no longer part of a `guaranteeCase` block, which _should_ be correct in this specific case since cancelation shouldn't happen on that fiber, but it's hard to reason about, and also the `onOutcome` behaviour wrt interruption seems to have changed, because now if `oc` is `Canceled` `onOutcome` is no longer called on it.

Both these changes seem orthogonal to the actual bug fix, are they there on purpose?

Reply:
So, I played with this a bit, and this works:
```scala
.guaranteeCase { oc =>                                       
  lease.cancel.rethrow.guaranteeCase { leaseOutcome =>       
    val propagateOutcome = leaseOutcome match {              
      case Outcome.Succeeded(fu) =>                          
        onOutcome(oc <* Outcome.succeeded(fu), Either.unit)  
      case _ =>                                              
        F.unit                                               
    }                                                        
                                                             
    propagateOutcome >> available.release >> decrementRunning
  }.handleError(_ => ())                                     
}.handleError(_ => ())                                       
```

The suggestion I meant to offer was only to move `available.release >> decrementRunning` inside the `guaranteeCase`, but as it turns out you actually don't need the `Errored` case when propagating results. 
I find that behaviour pretty confusing, since `.flatMap` short circuits but `guaranteeCase` has a `Succeeded` result (with a failed EitherT inside).
As an aside, the two `handleError(_ => ())` aren't actually needed but not having them might result in stacktraces getting logged

Reply:
In any case, it's actually easier to think about that `forceR` inside the `guaranteeCase`, I'll add a commit and merge